lib/diskoverflow/sorted.go

```diff
@@ -0,0 +1,315 @@
+// Copyright (C) 2018 The Syncthing Authors.
+//
+// This Source Code Form is subject to the terms of the Mozilla Public
+// License, v. 2.0. If a copy of the MPL was not distributed with this file,
+// You can obtain one at https://mozilla.org/MPL/2.0/.
+
+package diskoverflow
+
+import (
+	"bytes"
+	"encoding/binary"
+	"fmt"
+	"sort"
+)
+
+const suffixLength = 8
+
+// Sorted is a disk-spilling container, that sorts added values by the
+// accompanying keys.
+// You must not add new values after calling PopFirst/-Last.
+type Sorted interface {
+	Add(k []byte, v Value)
+	PopFirst(v Value) bool
+	PopLast(v Value) bool
+	NewIterator(reverse bool) Iterator
+	Bytes() int
+	Items() int
+	SetOverflowBytes(bytes int)
+	Close()
+}
+
+type sorted struct {
+	commonSorted
+	base
+}
+
+type keyIterator interface {
+	Iterator
+	key() []byte
+}
+
+type commonSorted interface {
+	common
+	add(k []byte, v Value)
+	getFirst(v Value) bool
+	getLast(v Value) bool
+	dropFirst(bytes int)
+	dropLast(bytes int)
+	newIterator(p iteratorParent, reverse bool) keyIterator
+}
+
+// NewSorted returns an implementaiton of Sorted, spilling to disk at location.
+func NewSorted(location string) Sorted {
+	o := &sorted{base: newBase(location)}
+	o.commonSorted = &memorySorted{}
+	return o
+}
+
+func (o *sorted) Add(k []byte, v Value) {
+	if o.startSpilling(o.Bytes() + v.Bytes()) {
+		d := v.Marshal()
+		newSorted := newDiskSorted(o.location)
+		if o.iterating {
+			panic(concurrencyMsg)
+		}
+		o.iterating = true
+		it := o.newIterator(o, false)
+		for it.Next() {
+			v.Reset()
+			it.Value(v)
+			newSorted.add(it.key(), v)
+		}
+		it.Release()
+		o.commonSorted.Close()
+		o.commonSorted = newSorted
+		o.spilling = true
+		v.Reset()
+		v.Unmarshal(d)
+	}
+	o.add(k, v)
+}
+
+func (o *sorted) NewIterator(reverse bool) Iterator {
+	if o.iterating {
+		panic(concurrencyMsg)
+	}
+	o.iterating = true
+	return o.newIterator(o, reverse)
+}
+
+func (o *sorted) PopFirst(v Value) bool {
```

Comment:
I suspect it might be worth it to push the implementations of these downwards to the mem/db case, to avoid the db having to create and destroy iterators, read keys and values, etc twice (get+drop) in both cases.